File: productAppBE/src/main/java/com/hellokoding/springboot/restful/controller/ArticleAPI.java
package com.hellokoding.springboot.restful.controller;

import com.hellokoding.springboot.restful.model.Article;
import com.hellokoding.springboot.restful.model.ConnectionType;
import com.hellokoding.springboot.restful.model.dto.ArticleDto;
import com.hellokoding.springboot.restful.model.dto.IdContentDto;
import com.hellokoding.springboot.restful.model.dto.ItemConnectionDto;
import com.hellokoding.springboot.restful.model.dto.NameConnectionDto;
import com.hellokoding.springboot.restful.service.ArticleService;
import com.hellokoding.springboot.restful.service.ConnectionTypeService;
import com.hellokoding.springboot.restful.service.attachments.AttachmentService;
import com.hellokoding.springboot.restful.service.dto.AttachmentDTO;
import com.hellokoding.springboot.restful.service.dto.EntityType;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.minidev.json.JSONUtil;
import org.springframework.core.io.InputStreamResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import javax.validation.Valid;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.sql.Blob;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.util.Date;
import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1/article")
@Slf4j
@RequiredArgsConstructor

public class ArticleAPI {

    private final ArticleService articleService;
    private final AttachmentService attachmentService;
    private final ConnectionTypeService ctypeService;

    //api/v1/article/search?title=title&hash=hash
    @GetMapping("/filter")
    public ResponseEntity<List<ArticleDto>> filter(@RequestParam(name = "title", required = false) List<String> title,
                                                   @RequestParam(name = "hash", required = false) List<String> hash,
                                                   @RequestParam(name = "author", required = false) List<String> author,
                                                   @RequestParam(name = "org", required = false) List<String> org,
                                                   @RequestParam(name = "location", required = false) List<String> location,
                                                   @RequestParam(name = "language", required = false) List<String> lang,
                                                   @RequestParam(name = "description", required = false) String descr,
                                                   @RequestParam(name = "text", required = false) String text,
                                                   @RequestParam(name = "miscellany", required = false) List<String> misc,
                                                   @RequestParam(name = "status", required = false) List<Integer> status,
                                                   @RequestParam(name = "startDate", required = false) String startDate,
                                                   @RequestParam(name = "endDate", required = false) String endDate,
                                                   @RequestParam(name = "mov", required = false) Integer movement) throws ParseException {

        List<ArticleDto> searchResult = articleService.filter(title, hash, author, org, location, lang, descr, text, misc, status, startDate, endDate, movement);
        return ResponseEntity.ok(searchResult);
    }

    @GetMapping("/search")
    public ResponseEntity<List<ArticleDto>> search(@RequestParam(name = "description", required = false) String descr,
                                                   @RequestParam(name = "text", required = false) String text,
                                                   @RequestParam(name = "status", required = false) List<Integer> status,
                                                   @RequestParam(name = "startDate", required = false) String startDate,
                                                   @RequestParam(name = "endDate", required = false) String endDate,
                                                   @RequestParam(name = "mov", required = false) Integer movement) throws ParseException {

        List<ArticleDto> searchResult = articleService.search(descr, text, status, startDate, endDate, movement);
        return ResponseEntity.ok(searchResult);
    }

    @GetMapping
    public ResponseEntity<List<ArticleDto>> findAll() {
        List<ArticleDto> all = articleService.findAll();
        return ResponseEntity.ok(all);
    }

    @GetMapping("/searchMaterial")
    public ResponseEntity<List<IdContentDto>> searchMaterial(@RequestParam(name = "q", required = true) String q,
                                                             @RequestParam(name = "mov", required = true) Integer mov) {
        List<IdContentDto> search = articleService.searchMaterial(q, mov);
        return ResponseEntity.ok(search);
    }

    @GetMapping("/searchMaterialById")
    public ResponseEntity<List<IdContentDto>> searchMaterialById(@RequestParam(name = "id", required = true) String id) {
        Integer intId = Integer.parseInt(id);
        List<IdContentDto> search = articleService.searchMaterialById(intId);
        return ResponseEntity.ok(search);
    }

    @PostMapping("/ids")
    public ResponseEntity<List<IdContentDto>> getMaterialsByIds(@Valid @RequestBody List<Integer> idList) {
        return ResponseEntity.ok(articleService.findByIds(idList));
    }

    @PostMapping("/symmids/{id}")
    public ResponseEntity<List<NameConnectionDto>> getMaterialsByIdsAndSymmetrically(@PathVariable Integer id, @Valid @RequestBody List<Integer> idList) {
        //List<IdContentDto> rrr = orgService.findByIdsAndSymmetrically(idList, id);
        return ResponseEntity.ok(articleService.findByIdsAndSymmetrically(idList, id));
    }

    @PostMapping("/connectionTypes")
    public ResponseEntity<List<ConnectionType>> getConnectionTypes() {
        return ResponseEntity.ok(ctypeService.findAll());
    }

    @PostMapping
    public ResponseEntity<Article> create(@Valid @RequestBody ArticleDto article) {
        Article art = articleService.save(article);

        System.out.println("SAVED ARTICLE");
        System.out.println(art);

        return ResponseEntity.ok(art);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ArticleDto> findById(@PathVariable Integer id) {
        Optional<ArticleDto> stock = articleService.findById(id);
        if (!stock.isPresent()) {
            log.error("Id " + id + " is not existed");
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok(stock.get());
    }

    @PutMapping("/{id}")
    public ResponseEntity<Article> update(@PathVariable Integer id, @Valid @RequestBody ArticleDto article) {
        article.setId(id);
        if (!articleService.findById(id).isPresent()) {
            log.error("Id " + id + " is not existed");
            ResponseEntity.badRequest().build();
        }
        return ResponseEntity.ok(articleService.save(article));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity delete(@PathVariable Integer id) {
        if (!articleService.findById(id).isPresent()) {
            log.error("Id " + id + " is not existed");
            ResponseEntity.badRequest().build();
        }
        articleService.deleteById(id);
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/attachment")
    public ResponseEntity deleteAttachment(@RequestParam(name = "entityId", required = true) Integer entityId,
                                           @RequestParam(name = "fileId", required = true) Integer fileId,
                                           @RequestParam(name = "fileName", required = true) String fileName) {

        boolean res = attachmentService.deleteAttachment(EntityType.ARTICLE, entityId, fileId, fileName);
        return ResponseEntity.ok(res);
    }

    @PostMapping("/attachment/{id}")
    public ResponseEntity<Integer> handleFileUpload(@PathVariable Integer id, @RequestParam("file") MultipartFile file) throws IOException {
        Integer attachmentId = attachmentService.createAttachment(EntityType.ARTICLE, id, Instant.now(), "admin", file.getOriginalFilename(), file.getBytes());
//        if (attachmentId == 0) return ResponseEntity.ok(attachmentId); //todo?
        return ResponseEntity.ok(attachmentId);
    }

    @GetMapping("/attachments/{id}")
    public ResponseEntity<List<AttachmentDTO>> getAttachments(@PathVariable Integer id) {
        List<AttachmentDTO> all = attachmentService.getAttachments(EntityType.ARTICLE, id);
        return ResponseEntity.ok(all);
    }

    public Optional<String> getExtensionByStringHandling(String filename) {
        return Optional.ofNullable(filename)
                .filter(f -> f.contains("."))
                .map(f -> f.substring(filename.lastIndexOf(".") + 1));
    }

    @GetMapping("/attachment")
    public ResponseEntity<InputStreamResource> getAttachment(@RequestParam(name = "entityId", required = true) Integer entityId, @RequestParam(name = "id", required = true) Integer id) {

        HttpHeaders headers = new HttpHeaders();
        ByteArrayInputStream bis = null;
        AttachmentDTO attachmentDTO = attachmentService.getAttachment(EntityType.ARTICLE, entityId, id);
        MediaType type = MediaType.APPLICATION_OCTET_STREAM;  //APPLICATION_PROBLEM_JSON;
        Optional<String> extension;

        try{
//            File uFile = null;
//            uFile = maintainFileService.getDocument(123456L);
            bis = new ByteArrayInputStream(attachmentDTO.getContent());
            String FN =  attachmentDTO.getName().replaceAll(",", ".");
            extension = getExtensionByStringHandling(FN);

            String headerView = "inline; filename=" + FN;
            headers.add("Content-Disposition", headerView);

            if (extension.isPresent()) {
                switch (extension.get().toLowerCase()) {
                    case "pdf": {
                        type = MediaType.APPLICATION_PDF;
                        break;
                    }
                    case "png": {
                        type = MediaType.IMAGE_PNG;
                        break;
                    }
                    case "jpg":
                    case "jpeg": {
                        type = MediaType.IMAGE_JPEG;
                        break;
                    }
                    case "gif": {
                        type = MediaType.IMAGE_GIF;
                        break;
                    }
                    case "txt": {
                        type = MediaType.TEXT_PLAIN;
                        break;
                    }

                    case "fb2":
                    case "htm":
                    case "html":{
                        type = MediaType.TEXT_HTML;
                        break;
                    }
                }
            }
        }
        catch(Exception e){
            System.out.printf("Exception", e);
        }

        return ResponseEntity
                .ok()
                .headers(headers)
                .contentType(type)
                .body(new InputStreamResource(bis));


//    public ResponseEntity<Blob> getAttachment(@RequestParam(name = "entityId", required = true) Integer entityId, @RequestParam(name = "id", required = true) Integer id) {
//    public ResponseEntity<byte[]> getAttachment(@RequestParam(name = "entityId", required = true) Integer entityId, @RequestParam(name = "id", required = true) Integer id) {

//        Blob attachment = attachmentService.getAttachment(EntityType.ARTICLE, entityId, id);
//        byte[] attachment = attachmentService.getAttachment(EntityType.ARTICLE, entityId, id);
//        return ResponseEntity.ok(attachment);
    }

    @GetMapping("/downloadAttachment")
    public ResponseEntity<InputStreamResource> downloadAttachment(@RequestParam(name = "entityId", required = true) Integer entityId, @RequestParam(name = "id", required = true) Integer id) {

        HttpHeaders headers = new HttpHeaders();
        ByteArrayInputStream bis = null;
        AttachmentDTO attachmentDTO = attachmentService.getAttachment(EntityType.ARTICLE, entityId, id);
//        MediaType type = MediaType.APPLICATION_OCTET_STREAM;  //APPLICATION_PROBLEM_JSON;

        try{
            bis = new ByteArrayInputStream(attachmentDTO.getContent());
            String FN =  attachmentDTO.getName().replaceAll(",", ".");

            String headerView = "attachment; filename=" + FN;
            headers.add("Content-Disposition", headerView);
//            headers.add("Content-Description", "File Transfer");
        }
        catch(Exception e){
            System.out.printf("Exception", e);
        }
        return ResponseEntity
                .ok()
                .headers(headers)
                .contentType(MediaType.APPLICATION_OCTET_STREAM)
                .body(new InputStreamResource(bis));
    }
}

File: frontend/src/components/article/article-api.js
import axios from 'axios'
import router from "../../router";

const AXIOS = axios.create({
    baseURL: `/api/v1/`,
    timeout: 10000
});


export default {
    // getAllArticles() {   //todo maybe in other views
    //     return AXIOS.get(`/article`);
    // },

    // getAllLanguages() {
    //     return AXIOS.get(`/language`);
    // },

    // getAllMovements() {
    //     return AXIOS.get(`/movement`);
    // },

    // getAllAuthors() {
    //     return AXIOS.get(`/person`);
    // },

    // getAuthorById(id, fn) {
    //     AXIOS
    //         .get(`/person/` + id)
    //         .then(response => fn(response))
    //         .catch(error => console.log(error))
    //     // return AXIOS.get(`/person/`+ id);
    // },

    create(article, fn) {
        AXIOS
            .post('/article', article)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.code === 401) {
                    this.error401Handling();
                }

            })
    },

    findById(id, fn) {
        AXIOS
            .get(`/article/` + id)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    getMaterialsByIds(idList, fn) {
        AXIOS
            .post(`/article/ids`, idList)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    getMaterialsByIdsAndSymmetrically(id,fn) {
        AXIOS
            .post(`/article/symmids/` + id)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    getConnectionTypes(fn) {
        AXIOS
            .post(`/article/connectionTypes`)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },


    update(id, article, fn) {
        AXIOS
            .put('/article/' + id, article)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    updateColor(id, article, fn) {
        AXIOS
            .put('/article/color/' + id, article)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    delete(id, fn) {
        AXIOS
            .delete('/article/' + id)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },


    searchMaterial(searchKey, fn) {
        AXIOS.get(
            `/article/searchMaterial?q=` + encodeURIComponent(searchKey))
            .then((response) => {
                if (response.status >= 200 && response.status < 300) {
                    return response;

                } else {
                    let error = new Error(response.statusText);
                    error.response = response;
                    throw error
                }
            })
            .then((response) => {
                // if (response.headers['content-type'] !== 'application/json') {
                //     let error = new Error('Некорректный ответ от сервера');
                //     error.response = response;
                //     throw error
                // }
                return response.data;

            }).then((json) => {
            fn(json);

        }).catch((error) => {
            console.log(error);
            if (error.response.status === 401) {
                this.error401Handling();
            }
        })
    },

    searchMaterialById(id, fn) {  //for the case with article creation with parent/child/equal
        AXIOS.get(
            `/article/searchMaterialById?id=` + encodeURIComponent(id))
            .then((response) => {
                if (response.status >= 200 && response.status < 300) {
                    return response;

                } else {
                    let error = new Error(response.statusText);
                    error.response = response;
                    throw error
                }
            }).then((response) => {
            // if (response.headers['content-type'] !== 'application/json') {
            //     let error = new Error('Некорректный ответ от сервера');
            //     error.response = response;
            //     throw error
            // }
            return response.data;

        }).then((json) => {
            fn(json);

        }).catch((error) => {
            console.log(error);
            if (error.response.status === 401) {
                this.error401Handling();
            }
        })
    },

    // filter(searchItem, searchKey, status, start, end, fn) {
    //     AXIOS.get(`article/search?` + searchItem + `=` + searchKey + `status=` + status + `&startDate=` + start + `&endDate=` + end)
    //         .then(response => fn(response))
    //         .catch(error => {
    //             console.log(error);
    //             if (error.response.status === 401) {
    //                 router.push('/login');
    //             }
    //         });
    // },

    filterAll(body, status, start, end, mov, fn) {
        AXIOS.get(`article/filter?` + body + `status=` + status + `&startDate=` + start + `&endDate=` + end + `&mov=` + mov)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    // searchHash(searchKey, status, start, end, fn) {
    //     AXIOS.get(`article/search?hash=` + encodeURIComponent(searchKey) + `&status=` + status + `&startDate=` + start + `&endDate=` + end)
    //         .then(response => fn(response))
    //         .catch(error => {
    //             console.log(error);
    //             if (error.response.status === 401) {
    //                 router.push('/login');
    //             }
    //         });
    // },
    //
    // searchAuthor(searchKey, status, start, end, fn) {
    //     AXIOS.get(`article/search?author=` + encodeURIComponent(searchKey) + `&status=` + status + `&startDate=` + start + `&endDate=` + end)
    //         .then(response => fn(response))
    //         .catch(error => {
    //             console.log(error);
    //             if (error.response.status === 401) {
    //                 router.push('/login');
    //             }
    //         });
    // },
    //
    // searchLanguage(searchKey, status, start, end, fn) {
    //     AXIOS.get(`article/search?language=` + encodeURIComponent(searchKey) + `&status=` + status + `&startDate=` + start + `&endDate=` + end)
    //         .then(response => fn(response))
    //         .catch(error => {
    //             console.log(error);
    //             if (error.response.status === 401) {
    //                 router.push('/login');
    //             }
    //         });
    // },
    //
    searchText(searchKey, status, start, end, mov, fn) {
        AXIOS.get(`article/search?text=` + encodeURIComponent(searchKey) + `&status=` + status + `&startDate=` + start + `&endDate=` + end + `&mov=` + mov)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    searchDescription(searchKey, status, start, end, mov, fn) {
        AXIOS.get(`article/search?description=` + encodeURIComponent(searchKey) + `&status=` + status + `&startDate=` + start + `&endDate=` + end + `&mov=` + mov)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            });
    },

    searchPeriodAndStatus(status, start, end, mov, fn) {
        AXIOS.get(`article/filter?status=` + status + `&startDate=` + start + `&endDate=` + end + `&mov=` + mov)
            .then(response => fn(response))
            .catch(error => {
                console.log(error);
                if (error.response.status === 401) {
                    this.error401Handling();
                }
            })
    },

    error401Handling() {
        localStorage.clear();
        router.push('/login');
    },
}


